pkg/sources/filesystem/filesystem.go

```diff
@@ -127,14 +127,25 @@ func (s *Source) scanDir(ctx context.Context, path string, chunksChan chan *sour
 			ctx.Logger().Error(err, "error walking directory")
 			return nil
 		}
+
 		fullPath := filepath.Join(path, relativePath)
 
+		// skip excluded directories early
+		if d.IsDir() {
+			if s.filter != nil && s.filter.ExcludeDir(fullPath) {
+				return fs.SkipDir
```

Comment:
Kinda random, but you can do the nil check inside the method to avoid repeating `filter != nil` in multiple places.

https://go.dev/play/p/GbvkbdAQTQd